internal/api/mfa.go

```diff
@@ -196,6 +199,42 @@ func (a *API) ChallengeFactor(w http.ResponseWriter, r *http.Request) error {
 	})
 }
 
+func (a *API) invokeHook(ctx context.Context, input any, output hooks.HookOutput) (hooks.HookOutput, *hooks.AuthHookError) {
+	var response []byte
+	switch input.(type) {
+	case hooks.MFAVerificationAttemptInput:
+		payload, err := json.Marshal(&input)
+		if err != nil {
+			panic(err)
+		}
+
+		if err := a.db.Transaction(func(tx *storage.Connection) error {
+			// We rely on Postgres timeouts to ensure the function doesn't overrun
+			timeoutQuery := tx.RawQuery(fmt.Sprintf("set local statement_timeout TO '%d';", hooks.DefaultTimeout))
+			if terr := timeoutQuery.Exec(); terr != nil {
+				return terr
+			}
+			query := tx.RawQuery(fmt.Sprintf("SELECT * from %s(?)", a.config.Hook.MFAVerificationAttempt.HookName), payload)
+			terr := query.First(&response)
+			if terr != nil {
+				return terr
+			}
+			return nil
+		}); err != nil {
+			return nil, hooks.HookError(err.Error())
+		}
+		if err = json.Unmarshal(response, &output); err != nil {
+			return nil, hooks.HookError(err.Error())
+		}
+		if output.IsError() {
+			return nil, hooks.HookError(output.Error())
+		}
```

Comment:
I prefer not to wrap errors unnecessarily. If there's a problem with the JSON, just return the error from JSON. Wrapping them up can cause it to be very hard to check the types and react accordingly. 

Finally, the `Output` struct can be made to be an "error" by implementing the `error` interface. So no need to wrap yet another `HookError` struct -- it's not going to be used anywhere anyway.